File: templates/js/mavlink_frame.js
import { CrcX25 } from './crc.js';
import { MavlinkVersion } from './mavlink_version.js';

export class MavlinkFrame {
  static MAVLINK_STX_V1 = 0xfe;
  static MAVLINK_STX_V2 = 0xfd;

  constructor(version, sequence, systemId, componentId, message) {
    this.version = version;
    this.sequence = sequence;
    this.systemId = systemId;
    this.componentId = componentId;
    this.message = message;
  }

  static v1(sequence, systemId, componentId, message) {
    return new MavlinkFrame(MavlinkVersion.V1, sequence, systemId, componentId, message);
  }

  static v2(sequence, systemId, componentId, message) {
    return new MavlinkFrame(MavlinkVersion.V2, sequence, systemId, componentId, message);
  }

  serialize() {
    if (this.version === MavlinkVersion.V1) {
      return this._serializeV1();
    }
    return this._serializeV2();
  }

  _serializeV1() {
    const payload = this.message.serialize();
    const payloadLength = payload.length;
    const frame = new Uint8Array(8 + payloadLength);
    frame[0] = MavlinkFrame.MAVLINK_STX_V1;
    frame[1] = payloadLength;
    frame[2] = this.sequence;
    frame[3] = this.systemId;
    frame[4] = this.componentId;
    frame[5] = this.message.mavlinkMessageId;

    const crc = new CrcX25();
    crc.accumulate(payloadLength);
    crc.accumulate(this.sequence);
    crc.accumulate(this.systemId);
    crc.accumulate(this.componentId);
    crc.accumulate(this.message.mavlinkMessageId);

    for (let i = 0; i < payloadLength; i++) {
      frame[6 + i] = payload[i];
      crc.accumulate(payload[i]);
    }
    crc.accumulate(this.message.mavlinkCrcExtra);

    frame[frame.length - 2] = crc.crc & 0xff;
    frame[frame.length - 1] = (crc.crc >> 8) & 0xff;
    return frame;
  }

  _serializeV2() {
    const incompatibilityFlags = 0;
    const compatibilityFlags = 0;
    const payload = MavlinkFrame._trimTrailingZeros(this.message.serialize());
    const payloadLength = payload.length;
    const messageId = this.message.mavlinkMessageId;
    const messageIdBytes = [
      messageId & 0xff,
      (messageId >> 8) & 0xff,
      (messageId >> 16) & 0xff,
    ];

    const frame = new Uint8Array(12 + payloadLength);
    frame[0] = MavlinkFrame.MAVLINK_STX_V2;
    frame[1] = payloadLength;
    frame[2] = incompatibilityFlags;
    frame[3] = compatibilityFlags;
    frame[4] = this.sequence;
    frame[5] = this.systemId;
    frame[6] = this.componentId;
    frame[7] = messageIdBytes[0];
    frame[8] = messageIdBytes[1];
    frame[9] = messageIdBytes[2];

    const crc = new CrcX25();
    crc.accumulate(payloadLength);
    crc.accumulate(incompatibilityFlags);
    crc.accumulate(compatibilityFlags);
    crc.accumulate(this.sequence);
    crc.accumulate(this.systemId);
    crc.accumulate(this.componentId);
    for (const byte of messageIdBytes) {
      crc.accumulate(byte);
    }

    for (let i = 0; i < payloadLength; i++) {
      frame[10 + i] = payload[i];
      crc.accumulate(payload[i]);
    }
    crc.accumulate(this.message.mavlinkCrcExtra);

    frame[frame.length - 2] = crc.crc & 0xff;
    frame[frame.length - 1] = (crc.crc >> 8) & 0xff;
    return frame;
  }

  static _trimTrailingZeros(payload) {
    let trimmedLength = payload.length;
    while (trimmedLength > 0 && payload[trimmedLength - 1] === 0) {
      trimmedLength -= 1;
    }
    if (trimmedLength === payload.length) {
      return payload;
    }
    return payload.subarray(0, trimmedLength);
  }
}

File: templates/rust/mavlink_frame.rs
use crate::crc::CrcX25;
use crate::mavlink_message::MavlinkMessage;
use crate::mavlink_version::MavlinkVersion;

pub struct MavlinkFrame {
    pub version: MavlinkVersion,
    pub sequence: u8,
    pub system_id: u8,
    pub component_id: u8,
    pub message: Box<dyn MavlinkMessage>,
}

impl MavlinkFrame {
    pub const MAVLINK_STX_V1: u8 = 0xFE;
    pub const MAVLINK_STX_V2: u8 = 0xFD;

    pub fn v1(
        sequence: u8,
        system_id: u8,
        component_id: u8,
        message: Box<dyn MavlinkMessage>,
    ) -> Self {
        Self {
            version: MavlinkVersion::V1,
            sequence,
            system_id,
            component_id,
            message,
        }
    }

    pub fn v2(
        sequence: u8,
        system_id: u8,
        component_id: u8,
        message: Box<dyn MavlinkMessage>,
    ) -> Self {
        Self {
            version: MavlinkVersion::V2,
            sequence,
            system_id,
            component_id,
            message,
        }
    }

    pub fn serialize(&self) -> Vec<u8> {
        match self.version {
            MavlinkVersion::V1 => self.serialize_v1(),
            MavlinkVersion::V2 => self.serialize_v2(),
        }
    }

    fn serialize_v1(&self) -> Vec<u8> {
        let payload = self.message.serialize();
        let payload_length = payload.len();
        let mut frame = vec![0u8; 8 + payload_length];
        frame[0] = Self::MAVLINK_STX_V1;
        frame[1] = payload_length as u8;
        frame[2] = self.sequence;
        frame[3] = self.system_id;
        frame[4] = self.component_id;
        frame[5] = self.message.mavlink_message_id() as u8;

        let mut crc = CrcX25::new();
        crc.accumulate(frame[1]);
        crc.accumulate(frame[2]);
        crc.accumulate(frame[3]);
        crc.accumulate(frame[4]);
        crc.accumulate(frame[5]);

        for (i, byte) in payload.iter().enumerate() {
            frame[6 + i] = *byte;
            crc.accumulate(*byte);
        }
        crc.accumulate(self.message.mavlink_crc_extra());

        let checksum = crc.crc();
        let frame_len = frame.len();
        frame[frame_len - 2] = (checksum & 0xFF) as u8;
        frame[frame_len - 1] = (checksum >> 8) as u8;
        frame
    }

    fn serialize_v2(&self) -> Vec<u8> {
        let incompatibility_flags = 0u8;
        let compatibility_flags = 0u8;
        let payload = Self::trim_trailing_zeros(&self.message.serialize());
        let payload_length = payload.len();
        let message_id = self.message.mavlink_message_id();
        let message_id_bytes = [
            (message_id & 0xFF) as u8,
            ((message_id >> 8) & 0xFF) as u8,
            ((message_id >> 16) & 0xFF) as u8,
        ];

        let mut frame = vec![0u8; 12 + payload_length];
        frame[0] = Self::MAVLINK_STX_V2;
        frame[1] = payload_length as u8;
        frame[2] = incompatibility_flags;
        frame[3] = compatibility_flags;
        frame[4] = self.sequence;
        frame[5] = self.system_id;
        frame[6] = self.component_id;
        frame[7] = message_id_bytes[0];
        frame[8] = message_id_bytes[1];
        frame[9] = message_id_bytes[2];

        let mut crc = CrcX25::new();
        crc.accumulate(frame[1]);
        crc.accumulate(frame[2]);
        crc.accumulate(frame[3]);
        crc.accumulate(frame[4]);
        crc.accumulate(frame[5]);
        crc.accumulate(frame[6]);
        for byte in message_id_bytes {
            crc.accumulate(byte);
        }

        for (i, byte) in payload.iter().enumerate() {
            frame[10 + i] = *byte;
            crc.accumulate(*byte);
        }
        crc.accumulate(self.message.mavlink_crc_extra());

        let checksum = crc.crc();
        let frame_len = frame.len();
        frame[frame_len - 2] = (checksum & 0xFF) as u8;
        frame[frame_len - 1] = (checksum >> 8) as u8;
        frame
    }

    fn trim_trailing_zeros(payload: &[u8]) -> Vec<u8> {
        let mut trimmed_length = payload.len();
        while trimmed_length > 0 && payload[trimmed_length - 1] == 0 {
            trimmed_length -= 1;
        }
        payload[..trimmed_length].to_vec()
    }
}
